src/Common/GuardedPoolAllocator.h

```diff
@@ -0,0 +1,72 @@
+#include <mutex>
+
+#include <Common/GuardedPoolAllocatorState.h>
+
+namespace gwp_asan 
+{
+
+class GuardedPoolAllocator;
+
+ALWAYS_INLINE GuardedPoolAllocator * getAllocator();
+
+class GuardedPoolAllocator 
+{
+public:
+    constexpr GuardedPoolAllocator() {}
+    GuardedPoolAllocator(const GuardedPoolAllocator &) = delete;
+    GuardedPoolAllocator &operator=(const GuardedPoolAllocator &) = delete;
+
+    ~GuardedPoolAllocator() = default;
+
+    void init();
+
+    void * allocate(size_t bytes, size_t alignment);
+
+    void deallocate(void * ptr);
+
+    bool shouldSample();
+
+    ALWAYS_INLINE bool pointerIsMine(const void *ptr) const {
+        return state.pointerIsMine(ptr);
+    }
+
+private:
+    static constexpr size_t kInvalidSlotID = SIZE_MAX;
+
+    /// Map and unmap memory with mmap
+    void *map(size_t Size) const;
+
+    void * reserveGuardedPool(size_t size);
+
+    size_t reserveSlotInGuardedPool();
+
+    // allocateInGuardedPool() ptr and size must be a subrange of the previously
+    // reserved pool range.
+    void allocateInGuardedPool(void *ptr, size_t size) const;
+    // deallocateInGuardedPool() ptr and size must be an exact pair previously
+    // passed to allocateInGuardedPool().
+    void deallocateInGuardedPool(void *ptr, size_t size) const;
+
+    static uintptr_t alignUp(uintptr_t ptr, size_t alignment);
+    static uintptr_t alignDown(uintptr_t ptr, size_t alignment);
+
+    static size_t getPlatformPageSize();
+
+    /// Return random size_t from [0, upper_bound - 1]
+    size_t getRandomSizeT(size_t upper_bound) const;
+
+    /// Array of free slots and its size. When a slot is freed it is added
+    /// to this array. When a free slot is requested, a random element of this
+    /// array is chosen, swapped with last element and the array is shrinked.
+    size_t *freeSlots = nullptr;
+    size_t freeSlotsLength = 0;
+
+    size_t sampleRate = 2;
+
+    size_t currentlyAllocatedSlots = 0;
+    
+    AllocatorState state;
+    std::mutex poolReservationMutex;
```

Comment:
AFAICS it is possible to use `thread_local` over `mutex` here?

It is important to avoid mutex for allocation hot code path (yes this is not hot-path, but still if it is possible I think it worth doing so)

Reply:
This mutex is used for accessing shared state: array of free slots, so I think it will not be possible to use thread_local here.

Reply:
Have you considered making this array thread-local too?
In other words make memory pool per-thread.
But, this will increase memory fragmentation, so maybe it does not worth it...

Reply:
Yes, this will increase fragmentation + total size of pool will be less tunable i.e. it will be a multiple of number of threads.

As I see it, it will be a really cold path (we'll set default sample rate to something like `100000`), so it should not be a big problem.

I suggest we observe impact on performance, which it will be producing, and replace this with thread_local pools if results turn out to be poor.

Reply:
>I suggest we observe impact on performance, which it will be producing, and replace this with thread_local pools if results turn out to be poor.

OK.

>As I see it, it will be a really cold path (we'll set default sample rate to something like 100000), so it should not be a big problem.

I think that it worth adding some config directive to change this sampling ratio.